Initialized collections

narimiran opened the pull request:
A lot of code has been moved around so the diff looks bigger than it really is.

### Important differences

* hashcommon.nim (this is a new file):
    * `template rawGetKnownHCImpl` 
* setimpl.nim (this is a new file): 
    * `template initImpl`
    * `template rawInsertImpl`
    * `template inclImpl`
* sets.nim:
    * `proc isValid` is now deprecated
* tableimpl.nim:
    * `template maybeRehashPutImpl`
    * `template initImpl`
    * `template insertImpl`

alehander42 commented:
```nim
import tables

var a: Table[string, int]
a["a"] = 0

var b: Table[string, int] = initTable[string, int]()
b["a"] = 0

echo a, a.hasKey("a")
echo b, b.hasKey("a")
```

should be fixed: hasKey returns false for first

narimiran commented:
> should be fixed: hasKey returns false for first

@alehander42 thanks for the testing and raising my attention to it. This also revealed several other (but similar) bugs which I have fixed now. I've also added a lots of tests for uninitialized tables and sets, so hopefully these embarrassing bugs won't go (almost) unnoticed the next time.

kaushalmodi commented on lib/pure/collections/sets.nim:
```diff
@@ -489,7 +341,7 @@ proc pop*[A](s: var HashSet[A]): A =
     doAssertRaises(KeyError, echo s.pop)
 
-  for h in 0..high(s.data):
+  for h in 0 .. high(s.data):
```
I love these "space around *dotdot*" consistency fixes!

timotheecour commented:
@narimiran, while this fixes https://github.com/nim-lang/RFCs/issues/101 and makes API easier to use, was there any benchmark done with this PR to assess the performance impact?

refs https://github.com/nim-lang/RFCs/issues/101#issuecomment-418044191 before this PR someone did something similar and noticed it caused a performance impact (impacting each table access)

>> we just need to check if t.data.len == 0 before using the table in []=, [] etc. I did a small benchmark (inserting 1G elements). For 64-element tables, the change seems to make the whole benchmark about 4% slower.

> Gah, that's pretty bad, any ideas why?
